### PR TITLE
:sparkles: Add E2E test to verify linked clone

### LINKED ISSUE #1765: No test verifies VMs are created as linked clone
/kind feature

**Describe the solution you'd like**

An E2E test verifying linked clone should be added.

When `spec.cloneMode` of `VSphereMachine` resource is specified to `linkedClone`, a corresponding virtual machine has to be created by using linked clone. But there is no test that verifies this behavior, so it could breaks without you noticing. For instance, #1611 reported such a bug.

**Anything else you would like to add:**

**Environment:**

### PR DESCRIPTION
**Pull request by ntoofu**
**What this PR does / why we need it**:

This PR adds E2E test to verify that a virtual machine whose `cloneMode` is specified to `linkedClone` is created as linked clone.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1765

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/linked_clone_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		ctx,
+		vms,
+		client.InNamespace(ns),
+		client.MatchingLabels{
+			clusterv1.ClusterLabelName: clusterName,
+		})
+	Expect(err).ToNot(HaveOccurred())
+
+	return vms.Items
+}
```
Comment (srm09) [inline test/e2e/linked_clone_test.go:136]: Use the `getVSphereVMsForCluster` defined in the `storage_policy_test.go` file instead.

**Comment by ntoofu** [comment]
[In artifacts of failed CI test](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/1766/pull-cluster-api-provider-vsphere-full-e2e/1625020264309854208/artifacts/clusters/bootstrap/logs/capi-kubeadm-bootstrap-system/capi-kubeadm-bootstrap-controller-manager/capi-kubeadm-bootstrap-controller-manager-5cccf5d8cd-tsmhr/), I found logs shown below:
```
E0213 06:43:56.170856       1 kubeadmconfig_controller.go:312] "Error creating remote cluster client" err="failed to create client for Cluster dhcp-overrides-tgnous/dhcp-overrides-2rq32b: Get \"https://192.168.6.162:6443/api?timeout=10s\": x509: certificate signed by unknown authority (possibly because of \"crypto/rsa: verification error\" while trying to verify candidate authority certificate \"kubernetes\") - error from a previous attempt: read tcp 10.244.0.9:44150->192.168.6.162:6443: read: connection reset by peer" controller="kubeadmconfig" controllerGroup="bootstrap.cluster.x-k8s.io" controllerKind="KubeadmConfig" KubeadmConfig="dhcp-overrides-tgnous/dhcp-overrides-2rq32b-md-0-ll5cn" namespace="dhcp-overrides-tgnous" name="dhcp-overrides-2rq32b-md-0-ll5cn" reconcileID=b8e0018a-b886-4974-87ca-19731ec45360 Machine="dhcp-overrides-tgnous/dhcp-overrides-2rq32b-md-0-5c445dfdb-dm94k" Machine="dhcp-overrides-tgnous/dhcp-overrides-2rq32b-md-0-5c445dfdb-dm94k" resourceVersion="1967" Cluster="dhcp-overrides-tgnous/dhcp-overrides-2rq32b"
```

However, I'm not sure that the certificate verification error is the root cause.

**Comment by ntoofu** [comment]
The certificate verification error disappeared when I retried test. But another error [occurred](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/1766/pull-cluster-api-provider-vsphere-full-e2e/1626627661054873600) in newly added test:

> [FAIL] Cluster creation with linked clone mode [It] should create a cluster by using linked clone to create nodes 

By investigating [logs](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/1766/pull-cluster-api-provider-vsphere-full-e2e/1626627661054873600/artifacts/clusters/bootstrap/logs/capv-system/capv-controller-manager/capv-controller-manager-6d49dd8cd9-rqdnv/manager.log), I found CAPV controller cloned VMs in full clone mode.

> I0217 17:07:57.639185       1 clone.go:269] "capv-controller-manager/vspherevm-controller/quick-start-1w4jrn/quick-start-frr36n-w85fz/vcenter: cloning machine" namespace="quick-start-1w4jrn" name="quick-start-frr36n-w85fz" cloneType=fullClone

I suspect that a template VM in CI environment has no snapshot, but I don't know how to confirm it. How should I handle this?

**Comment by sbueringer** [comment]
I wonder if we should combine this with other tests. E.g. the storage policy test also seems to just create a cluster and then is doing some validation afterwards. 

If both of those are just creating a default cluster and then running some validation maybe we should do this after the running the quickstart spec in capv_quick_start_test.go?

@randomvariable Do you know if this is possible? (Just find some examples with after each in CAPA, but not sure when exactly the aftereach is executed vs the aftereach from core CAPI that deletes the cluster)

**Comment by randomvariable** [comment]
> Do you know if this is possible? 

Not with the quickstart. Unfortunately, much of the CAPI test suite is too restrictive for stuff like this. It would be good to introduce hooks in them to do additional tests, and pass things in like the machine logs grabber, etc...

In CAPA, these sorts of tests are covered in the unmanaged functional suite, and we do as much as possible to consolidate these tests into a single cluster creation.

https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/test/e2e/suites/unmanaged/unmanaged_functional_test.go

**Comment by sbueringer** [comment]
@randomvariable Thx for the context.

Should have rememberd. For quickstart we actually implemented that because we use it in core CAPI to verify ownerRefs (we plan the same for CAPV, will create an issue soon): https://github.com/kubernetes-sigs/cluster-api/blob/caaa74482b51fae777334cd7a29595da1c06481e/test/e2e/quick_start_test.go#L40-L50

I think we could use `PostMachinesProvisioned` as well to add additional validation

**Comment by chrischdi** [comment]
/close

@killianmuldoon will open a new PR and follow-up / link via #1765

**Pull request closed without merging**